### PR TITLE
Adding `SourceMatchController` class.

### PR DESCRIPTION
**Pull request by abtink**
This commit adds a new class `SourceMatchController` which
contains all (existing) logic to maintain the source match table
and control when/how to enable or disable the source matching
feature.

It also updates `Child` class by changing the access level of
member variables related to the source matching feature (e.g.,
number of indirect queued messages) to private to ensure that
they are only modified by the new `SourceMatchController` class
while providing simple getter methods for use by other classes.

The changes in this commit also address two corner-case issues
related to source address matching feature:

- First issue where we could possibly enable the source address
  matching after successfully adding a short address while still
  having remaining pending entries.

- Second issue where an entry would not be added properly from a
  `SetSrcMatchAsShort()` call for a child, if there were more
  than one message queued for the child.

### CODE REVIEW

File: src/core/thread/mesh_forwarder.cpp [hunk 1]
```diff
@@ -1043,12 +904,16 @@ ThreadError MeshForwarder::HandleFrameRequest(Mac::Frame &aFrame)
 
     assert(error == kThreadError_None);
 
-    // set FramePending if there are more queued messages for the child
     aFrame.GetDstAddr(macDest);
 
+    // Set `FramePending` if there are more queued messages (excluding
+    // the current one being sent out) for the child (note `> 1` check).
+    // The case where the current message requires fragmentation is
+    // already checked and handled in `SendFramenet()` method.
```
Comment (jwhui) [inline src/core/thread/mesh_forwarder.cpp]: typo: `SendFragment`

File: src/core/thread/src_match_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+{
+    ClearTable();
+}
+
+otInstance *SourceMatchController::GetInstance(void)
+{
+    return mMeshForwarder.GetInstance();
+}
+
+void SourceMatchController::IncreaseMessageCount(Child &aChild)
```
Comment (jwhui) [inline src/core/thread/src_match_controller.cpp]: nit: `IncreaseMessageCount` -> `IncrementMessageCount`

File: src/core/thread/src_match_controller.cpp [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+{
+    if (aChild.mQueuedMessageCount++ == 0)
+    {
+        AddEntry(aChild);
+    }
+}
+
+void SourceMatchController::DecreaseMessageCount(Child &aChild)
```
Comment (jwhui) [inline src/core/thread/src_match_controller.cpp]: nit: `DecreaseMessageCount` -> `DecrementMessageCount`

File: src/core/thread/topology.hpp [hunk 4]
```diff
@@ -145,9 +147,28 @@ class Child : public Neighbor
         uint8_t mAttachChallenge[Mle::ChallengeTlv::kMaxSize]; ///< The challenge value
     };
     uint8_t      mNetworkDataVersion;                  ///< Current Network Data version
-    uint16_t     mQueuedIndirectMessageCnt;            ///< Count of queued messages
-    bool         mAddSrcMatchEntryShort : 1;           ///< Indicates whether or not to force add short address
-    bool         mAddSrcMatchEntryPending : 1;         ///< Indicates whether or not pending to add
+
+public:
```
Comment (jwhui) [inline src/core/thread/topology.hpp]: nit: `public:` is redundant with above

File: src/core/thread/src_match_controller.hpp [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+private:
+    /**
+     * This method clears the source match table.
+     *
+     */
+    void ClearTable(void);
+
+    /**
+     * This method enables or disables the source matching..
```
Comment (jwhui) [inline src/core/thread/src_match_controller.hpp]: nit: double `..`

Also, indicate that when source matching is disabled, the Frame Pending bit is set in all IEEE 802.15.4 ACKs to IEEE 802.15.4 Data Request messages.

File: src/core/thread/mesh_forwarder.cpp (hunk 1 above)
Comment (abtink) [inline src/core/thread/mesh_forwarder.cpp]: Done!. Thanks.

File: src/core/thread/src_match_controller.cpp (hunk 2 above)
Comment (abtink) [inline src/core/thread/src_match_controller.cpp]: Done!

File: src/core/thread/src_match_controller.cpp (hunk 3 above)
Comment (abtink) [inline src/core/thread/src_match_controller.cpp]: Done

File: src/core/thread/src_match_controller.cpp [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+}
+
+void SourceMatchController::Enable(bool aEnable)
+{
+    mEnabled = aEnable;
+    otPlatRadioEnableSrcMatch(GetInstance(), mEnabled);
+    otLogDebgMac(GetInstance(), "SrcAddrMatch - %sabling", mEnabled ? "En" : "Dis");
+}
+
+
```
Comment (librasungirl) [inline src/core/thread/src_match_controller.cpp]: nit: remove the additional blank line?

Comment (abtink) [inline src/core/thread/src_match_controller.cpp]: Thanks... Will fix it...

**Pull request merged**